cmd/os-instrumented.go

```diff
@@ -134,6 +134,19 @@ func Rename(src, dst string) error {
 	return os.Rename(src, dst)
 }
 
+// PanRename captures time taken to call Link/Unlink
+func PanRename(src, dst string) error {
+	defer updateOSMetrics(osMetricRename, src, dst)()
+	if err := os.Remove(dst); err != nil && !os.IsNotExist(err) {
```

Comment:
I'm not sure what is more effective in terms of performance, but just throwing the idea that we can omit first ```os.Remove()``` and call it only if attempt to Link fails with EEXIST. This should reduce one syscall in happy path when target file does not exist:

```
err = link(src, dst)
if err == EEXIST:
    err = remove(dst)
    if err == nil:
        err = link(src, dst)
```

However, it would make one additional syscall on re-uploads (failed first Link()), so it's debatable whether it's reasonable or not.

Reply:
It's some kind of fortune telling on coffee thick without any statistics. We need to know about common case: whether file exists or not. Until that, less code is more preferable to my mind.

I see no metrics, exclude time of execution of common syscalls to fs. We can add our own metrics gathering system(not so easy and fast to implement) or add simple counting with periodical println as a temporary solution(and remove it with future commits). Do you think if it's useful or not?

Reply:
Everything depends on client's workload and applications that will call S3 interface. On the other hand, I'd limit our decision tree only for small objects case, as the impact of additional syscall per X writes for big object is neglectable. And I would say that creation of new files should me much more common than overwriting existing ones. And of course, this is the case for performance benchmarks, that always create new objects unless some specific pattern is specified. But again, it does not mean that overwriting of existing objects is impossible usecase. 

As for embedded profiling, I do not think that this is necessary. PanFS provides tons of tracing, performance/ops stats and monitoring tools. Plus we also have dtrace/bcc that can help us to get detailed view.